coverage_point_calculator/src/lib.rs

```diff
@@ -0,0 +1,933 @@
+#[derive(Debug, Clone)]
+pub struct RewardableRadio {
+    radio_type: RadioType,
+    speedtests: Speedtests,
+    location_trust_scores: LocationTrustScores,
+    radio_threshold: RadioThreshold,
+    covered_hexes: CoveredHexes,
+    boosted_hex_eligibility: BoostedHexStatus,
+}
+
+/// Output of calculating coverage points for a [RewardableRadio].
+///
+/// The only data included was used for calculating coverage points.
+///
+/// - If more than the allowed speedtests were provided, only the speedtests
+///   considered are included here.
+///
+/// - When a radio covers boosted hexes, [location_trust_scores] will contain a
+///   trust score _after_ the boosted hex restriction has been applied.
+///
+/// - When a radio is not eligible for boosted hex rewards, [covered_hexes] will
+///   have no boosted_multiplier values.
+///
+#[derive(Debug)]
+pub struct CoveragePoints {
+    /// Value used when calculating poc_reward
+    pub total_coverage_points: Decimal,
+    /// Coverage Points collected from each Covered Hex
+    pub hex_coverage_points: Decimal,
+    /// Location Trust Multiplier, maximum of 1
+    pub location_trust_multiplier: Decimal,
+    /// Speedtest Mulitplier, maximum of 1
+    pub speedtest_multiplier: Decimal,
+    // ---
+    pub radio_type: RadioType,
+    pub radio_threshold: RadioThreshold,
+    pub speedtests: Vec<Speedtest>,
+    pub location_trust_scores: Vec<LocationTrust>,
+    pub covered_hexes: Vec<CoveredHex>,
+    pub boosted_hex_eligibility: BoostedHexStatus,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("signal level {0:?} not allowed for {1:?}")]
+    InvalidSignalLevel(SignalLevel, RadioType),
+}
+
+pub fn calculate_coverage_points(radio: RewardableRadio) -> CoveragePoints {
+    let hex_coverage_points = radio.covered_hexes.calculated_coverage_points();
+    let location_trust_multiplier = radio.location_trust_scores.multiplier;
+    let speedtest_multiplier = radio.speedtests.multiplier;
+
+    let coverage_points = hex_coverage_points * location_trust_multiplier * speedtest_multiplier;
+    let total_coverage_points = coverage_points.round_dp_with_strategy(2, RoundingStrategy::ToZero);
+
+    CoveragePoints {
+        total_coverage_points,
+        hex_coverage_points,
+        location_trust_multiplier,
+        speedtest_multiplier,
+        // Radio information
+        radio_type: radio.radio_type,
+        radio_threshold: radio.radio_threshold,
+        speedtests: radio.speedtests.speedtests,
+        location_trust_scores: radio.location_trust_scores.trust_scores,
+        covered_hexes: radio.covered_hexes.hexes,
+        boosted_hex_eligibility: radio.boosted_hex_eligibility,
+    }
+}
+
+impl RewardableRadio {
+    pub fn new(
+        radio_type: RadioType,
+        speedtests: Vec<Speedtest>,
+        location_trust_scores: Vec<LocationTrust>,
+        radio_threshold: RadioThreshold,
+        covered_hexes: Vec<RankedCoverage>,
```

Comment:
naming this covered_hexes is somewhat confusing when we have a Struct named CoveredHexes and then a field in the returned struct of the same name and type, maybe radio_hexes or just ranked_coverage is a better fit?

Reply:
Renamed to `ranked_coverage`, good catch.